### ISSUE TITLE
Secured WebSocket

### ISSUE THREAD

**Issue by grahamehorner** [aspnet/SignalR#451]
Please consider using WSS over WS as default, also allow client/server side certificate validation hooks; so that SignalR client connection may be revoked/expired via certificate expiry; prior to connection/execution authentication/authorisation on hub/method.

**Comment by davidfowl** [comment]
I'm not sure we can default to that anywhere, it would require a lot more infrastructure that isn't trivially set up by default. That decision needs to be made at the ASP.NET level before we can force everyone into it.

**Comment by grahamehorner** [comment]
@davidfowl understood; could it be an option for the client to inform the user that the connection is insecure ? 

IMO if a SignalR client uses WS instead of WSS that may be unaware the data is insecure? an alternative could be to use the Authorise attribute to forces the connection to be secured, the a RequiresSecuredConnection = true; which verifies that the endpoint(s) that the hub/method is using for the activation is secured.

**Comment by davidfowl** [comment]
@grahamehorner I don't see why that would be SignalR specific.

**Comment by grahamehorner** [comment]
@davidfowl It's more a transport specific, IMO the authorise attribute needs to verify the transport is secure like and as SignalR hubs can be exposed on multiple transports; IMO Authroisation needs to two things:-

1. the transport uses is secure ?
2. the claims/roles of the calling party?

I'm not sure how/if the hub at this point knows anything about the transport or can access information about the transport, however the process of activation/calling the hub is known by SignalR and therefore the attribute can be used to verify secured transports and/or claim/roles prior to activation.

above is just thinking aloud; 

@davidfowl your the man with the plan ;) 

BTW I'd the //build demo coding on stage with preview release BRAVE MAN!! cool #SocialWeather demo; I've been using #ESP8266 with Arduino for a few years now :D with a community built Azure library :D

**Comment by davidfowl** [comment]
> the transport uses is secure ?
> the claims/roles of the calling party?

@grahamehorner I think those 2 things are completely separate.

**Comment by grahamehorner** [comment]
@davidfowl so do you envision [RequireSecuredTransport()] as a attribute which can be applied to SignalR hubs/methods as well as MVC controlles/actions ?

**Comment by davidfowl** [comment]
@grahamehorner It's something that could exist at the middleware or EndPoint/Hub level, it's very similar to `[RequireHttps]` (if not the exact same thing).

**Comment by grahamehorner** [comment]
@davidfowl please correct me  if my understanding is wrong; but RequireHttps doesn't cover the use case when a is explicitly exposed on other transports ? want if a hub isn't published on a http/https endpoint ?

**Comment by davidfowl** [comment]
@grahamehorner Not sure what you mean. I was merely comparing the similarities between the 2 features.

**Comment by grahamehorner** [comment]
@davidfowl ok; I missed understood sometimes it's difficult for me to understand context/intent of email/chat; my issue sorry.

**Comment by grahamehorner** [comment]
@davidfowl Q: am I correct in thinking a SignalR hub can be exposed explicitly on transports without the need for HTTP/HTTP for example the hub only supporting WS/WSS or TCP etc ?

**Comment by davidfowl** [comment]
That's what we'd like to enable but it's not possible today.

**Comment by anurse** [comment]
It's definitely possible right now to limit you to using only WebSockets, though we haven't fully exposed it through Hubs (it's possible at the lower EndPoints layer and we intend to expose it all the way up). Right now though, we don't have a way to limit you to a Secured WebSocket though. I'm filing an issue to discuss that.

**Comment by moozzyk** [comment]
Closing in favour #457
